storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTaskTest.java

```diff
@@ -0,0 +1,414 @@
+        final TopicIdPartition tpId0 = new TopicIdPartition(topicId, new TopicPartition("sample", 0));
+        final TopicIdPartition tpId1 = new TopicIdPartition(topicId, new TopicPartition("sample", 1));
+        final TopicIdPartition tpId2 = new TopicIdPartition(topicId, new TopicPartition("sample", 2));
+        assertEquals(partitioner.metadataPartition(tpId0), partitioner.metadataPartition(tpId1));
+        assertEquals(partitioner.metadataPartition(tpId0), partitioner.metadataPartition(tpId2));
+
+        final int metadataPartition = partitioner.metadataPartition(tpId0);
+        consumer.updateEndOffsets(Collections.singletonMap(toRemoteLogPartition(metadataPartition), 0L));
+        final Set<TopicIdPartition> assignments = Collections.singleton(tpId0);
+        consumerTask.addAssignmentsForPartitions(assignments);
+        thread.start();
+        TestUtils.waitForCondition(() -> consumerTask.isUserPartitionAssigned(tpId0), "Timed out waiting for " + tpId0 + " to be assigned");
+
+        addRecord(consumer, metadataPartition, tpId0, 0);
+        addRecord(consumer, metadataPartition, tpId0, 1);
+        TestUtils.waitForCondition(() -> consumerTask.readOffsetForMetadataPartition(metadataPartition).equals(Optional.of(1L)), "Couldn't read record");
+        assertEquals(2, handler.metadataCounter);
+
+        // should only read the tpId1 records
+        consumerTask.addAssignmentsForPartitions(Collections.singleton(tpId1));
+        TestUtils.waitForCondition(() -> consumerTask.isUserPartitionAssigned(tpId1), "Timed out waiting for " + tpId1 + " to be assigned");
+        addRecord(consumer, metadataPartition, tpId1, 2);
+        TestUtils.waitForCondition(() -> consumerTask.readOffsetForMetadataPartition(metadataPartition).equals(Optional.of(2L)), "Couldn't read record");
+        assertEquals(3, handler.metadataCounter);
+
+        // shouldn't read tpId2 records because it's not assigned
+        addRecord(consumer, metadataPartition, tpId2, 3);
+        TestUtils.waitForCondition(() -> consumerTask.readOffsetForMetadataPartition(metadataPartition).equals(Optional.of(3L)), "Couldn't read record");
+        assertEquals(3, handler.metadataCounter);
+    }
+
+    @Test
+    public void testMaybeMarkUserPartitionsAsReady() throws InterruptedException {
+        final TopicIdPartition tpId = getIdPartitions("hello", 1).get(0);
+        final int metadataPartition = partitioner.metadataPartition(tpId);
+        consumer.updateEndOffsets(Collections.singletonMap(toRemoteLogPartition(metadataPartition), 2L));
+        consumerTask.addAssignmentsForPartitions(Collections.singleton(tpId));
+        thread.start();
+
+        TestUtils.waitForCondition(() -> consumerTask.isUserPartitionAssigned(tpId), "Waiting for " + tpId + " to be assigned");
+        assertTrue(consumerTask.isMetadataPartitionAssigned(metadataPartition));
+        assertFalse(handler.isPartitionInitialized.containsKey(tpId));
```

Comment:
Does this mean, when the partition contains no records, the partition is considered as uninitialized at this state? Does that make sense?

Reply:
The partition will get initialized if start and end offsets are same. It is handled in `maybeMarkUserPartitionsAsReady`

```
                    if (readOffset + 1 >= holder.endOffset || holder.endOffset.equals(holder.startOffset)) {
                        markInitialized(utp);
                    }
```

Reply:
Thanks for the explanation. So, my question is, at this stage, we've verified the user partition and metadata partition are assigned:
```
 TestUtils.waitForCondition(() -> consumerTask.isUserPartitionAssigned(tpId), "Waiting for " + tpId + " to be assigned");
 assertTrue(consumerTask.isMetadataPartitionAssigned(metadataPartition));
```
And now, the start and end offset should be the same (all are 0), so why does the partition cannot be initialized?
```
assertFalse(handler.isPartitionInitialized.containsKey(tpId));
```

Reply:
The start and end offsets are actually not the same. We have moved the endoffset to 2  on the third line of the method. If we did not do this step, the partition would have initialized.